### PR TITLE
Allow setting sort_field from model

### PR DESCRIPTION
**Pull request by remiberthoz**
It is now possible to define a `sort_field` value for notes by
setting a `sort_field_number` when creating the model.
When the `sort_field` argument is unused upon note creation, then
the field corresponding to the models `sort_field_number` will be
used. The default value is `1` (for first field).

The changes are compatible with previous versions.

### CODE REVIEW

File: genanki/note.py [hunk 1]
```diff
@@ -7,22 +7,17 @@ class Note:
...
+    else:
+        self.sort_field_value = sort_field  # overwrite model defined value if sort_field is passed
     self.tags = tags or []
     try:
       self.guid = guid
     except AttributeError:
       # guid was defined as a property
       pass
 
-  @property
```
Comment (kerrickstaley) [inline genanki/note.py]: Removing the `sort_field` property is a backwards-incompatible change. Could you rework the PR to preserve the `.sort_field` property?

Comment (remiberthoz) [reply genanki/note.py]: Sure yes! I did not realise this was breaking backward compatibility sorry. I never used python `@property` before. I'll force-push a commit, rebasing on top of your `master` branch and squashing the fix.

File: README.md [hunk 2]
```diff
@@ -122,12 +129,6 @@ class MyNote(genanki.Note):
     return genanki.guid_for(self.fields[0], self.fields[1])
 ```
 
-## sort_field
-Anki has a value for each `Note` called the `sort_field`. Anki uses this value to sort the cards in the Browse
-interface. Anki also is happier if you avoid having two notes with the same `sort_field`, although this isn't strictly
-necessary. By default, the `sort_field` is the first field, but you can change it by passing `sort_field=` to `Note()`
-or implementing `sort_field` as a property in a subclass (similar to `guid`).
-
 ## YAML for Templates (and Fields)
```
Comment (kerrickstaley) [inline README.md:132]: I'd prefer to keep this section here, and just add the following text.
```suggestion
You can also pass `sort_field_index=` to `Model()` to change the sort field. `0` means the first field in the Note, `1` means the second, etc.

## YAML for Templates (and Fields)
```

File: README.md [hunk 3]
```diff
@@ -31,6 +31,7 @@ my_model = genanki.Model(
     {'name': 'Question'},
     {'name': 'Answer'},
   ],
+  sort_field_number=2,
```
Comment (kerrickstaley) [inline README.md:34]: Let's keep the example as simple as possible and remove this line.

File: README.md [hunk 4]
```diff
@@ -48,6 +49,12 @@ You need to pass a `model_id` so that Anki can keep track of your model. It's im
 for each `Model` you define. Use `random.randrange(1 << 30, 1 << 31)` to generate a suitable model_id, and hardcode it
 into your `Model` definition.
 
+Finally, the notes browser inside Anki will sort cards with a special sorting field. The default value for this field
```
Comment (kerrickstaley) [inline README.md:52]: Per below comment, could you remove this section?

File: genanki/model.py [hunk 5]
```diff
@@ -8,13 +8,14 @@ class Model:
   FRONT_BACK = 0
   CLOZE = 1
 
-  def __init__(self, model_id=None, name=None, fields=None, templates=None, css='', model_type=FRONT_BACK):
+  def __init__(self, model_id=None, name=None, fields=None, templates=None, css='', model_type=FRONT_BACK, sort_field_number=1):
```
Comment (kerrickstaley) [inline genanki/model.py:11]: Let's rename this to `sort_field_index` and make it 0-based.

File: genanki/model.py [hunk 6]
```diff
@@ -8,13 +8,14 @@ class Model:
...
     self.model_id = model_id
     self.name = name
     self.set_fields(fields)
     self.set_templates(templates)
     self.css = css
     self.model_type = model_type
+    self.sort_field_id = sort_field_number-1
```
Comment (kerrickstaley) [inline genanki/model.py:18]: So this line becomes `self.sort_field_index = sort_field_index`.

Note that `self.sort_field_id = sort_field_number-1` is incompatible with a general API principle of genanki, which is that
```
c = Class(foo='bar')
```
usually does the same thing as
```
c = Class()
c.foo = 'bar'
```

File: genanki/model.py [hunk 7]
```diff
@@ -108,7 +109,7 @@ def to_json(self, now_ts, deck_id):
       "mod": now_ts,
       "name": self.name,
       "req": self._req,
-      "sortf": 0,
+      "sortf": self.sort_field_id,
```
Comment (kerrickstaley) [inline genanki/model.py:112]: This is interesting, I didn't notice that the sort field index was encoded in the model's JSON representation. I wonder what happens when you try to set `sort_field` on a `Note` but leave `Model.sort_field_index` as 0. In that case, Anki will have conflicting information on what the sort field actually is.

My guess is that `sortf` in the model JSON (in `col.models`) is only used when creating new notes, and `notes.sfld` is used for sorting the data in the browser. So this inconsistency won't break Anki, but it could cause confusion if someone creates a new note based on a model from a genanki deck. This requires further investigation and it may be best to change the API to disable setting `Note.sort_field`.

Another funny thing I noticed is that when Anki generates a deck, the `notes.sfld` column is of `integer` type but contains text, which is [apparently a thing SQLite allows you to do](https://dba.stackexchange.com/questions/106364/text-string-stored-in-sqlite-integer-column). We do the same thing in `genanki`.

**Review by kerrickstaley — CHANGES_REQUESTED** [review]

File: genanki/model.py (hunk 7 above)
Comment (remiberthoz) [reply genanki/model.py]: Indeed this requires investigation. I can test this if you wish, but I don't have a lot of time these days...

**Comment by remiberthoz** [comment]
Sorry I don't know why it says I closed the issue... I forced-push on my master branch after rebasing on top of your master branch and resolving conflicts. I am normally two commits ahead.

File: genanki/model.py (hunk 7 above)
Comment (kerrickstaley) [reply genanki/model.py]: Yeah, mostly just noting this down so I can maybe investigate it later.

**Comment by kerrickstaley** [comment]
@remiberthoz I pushed [one quick change to your PR](https://github.com/kerrickstaley/genanki/pull/42/commits/ad79fdc77594bb334ffa1b72a156b7128cbde827) to avoid potentially breaking API compatibility, and it looks good to merge :smile: Will merge it momentarily once the CI tests pass.

One note is that we need to add tests that exercise this functionality. I can do that in the next few days or if you want to take it on, just comment here to let me know so we don't duplicate work.

**Review by kerrickstaley — APPROVED** [review]

**Comment by remiberthoz** [comment]
Hey, I will write the tests!

What behaviors would you like to test against? Of course, that the sorting index actually sorts when its used. Do you have edge-cases in mind?

**Comment by kerrickstaley** [comment]
@remiberthoz I think a simple ["happy path" test case](https://en.wikipedia.org/wiki/Happy_path) would be fine for this feature, just to make sure nobody accidentally breaks it in the future.
